### PR TITLE
Fix windows ouput frame size

### PR DESCRIPTION
**Pull request by RohanPunjani**

### CODE REVIEW

File: scap/src/capturer/engine/win/mod.rs [hunk 1]
```diff
@@ -15,46 +24,110 @@ use windows_capture::{
...
+#[derive(Debug)]
 struct Capturer {
     pub tx: mpsc::Sender<Frame>,
+    pub crop: Option<CGRect>,
 }
 
 impl Capturer {
     pub fn new(tx: mpsc::Sender<Frame>) -> Self {
-        Capturer { tx }
+        println!("I am here inside impl_capturer_new");
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs]: Nit: Let's remove these unnecessary logs

File: scap/src/capturer/engine/win/mod.rs [hunk 2]
```diff
@@ -15,46 +24,110 @@ use windows_capture::{
...
 
 impl WindowsCaptureHandler for Capturer {
-    type Flags = mpsc::Sender<Frame>;
+    type Flags = FlagStruct;
     type Error = Box<dyn std::error::Error + Send + Sync>;
 
-    fn new(tx: Self::Flags) -> Result<Self, Self::Error> {
-        Ok(Self { tx })
+    fn new(flagValues: Self::Flags) -> Result<Self, Self::Error> {
+        println!("I am here inside WindowsCaptureHandler new");
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs]: This as well

File: scap/src/capturer/engine/win/mod.rs [hunk 3]
```diff
@@ -15,46 +24,110 @@ use windows_capture::{
...
+                    width: cropped_area.size.width as i32,
+                    height: cropped_area.size.height as i32,
+                    data: raw_frame_buffer.to_vec(),
+                };
+
+                self.tx.send(Frame::BGRA(bgr_frame))
+                    .expect("Failed to send data");
+            }
+            None => {
+                println!("Frame Arrived: {}x{}",
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs]: No need for this log here

File: scap/src/capturer/engine/win/mod.rs [hunk 4]
```diff
@@ -66,29 +139,115 @@ impl WindowsCaptureHandler for Capturer {
...
+    }
+
+    if output_width % 2 == 1 {
+        output_width -= 1;
+    }
+
+    if output_height % 2 == 1 {
+        output_height -= 1;
+    }
+    println!("Output frame size: [{}, {}]", output_width, output_height);
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs]: This as well

File: scap/src/capturer/engine/win/mod.rs [hunk 5]
```diff
@@ -15,46 +24,110 @@ use windows_capture::{
...
+                let start_x = cropped_area.origin.x as u32;
+                let start_y = cropped_area.origin.y as u32;
+                let end_x = (cropped_area.origin.x + cropped_area.size.width) as u32;
+                let end_y = (cropped_area.origin.y + cropped_area.size.height) as u32;
+
+                // crop the frame
+                let mut cropped_buffer = frame.buffer_crop(start_x, start_y, end_x, end_y)
+                    .expect("Failed to crop buffer");
+
+                println!("Frame Arrived: {}x{} and padding = {}",
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs]: And this

File: scap/src/main.rs [hunk 6]
```diff
@@ -62,22 +64,29 @@ fn main() {
...
                     i, frame.width, frame.height, frame.display_time
                 );
             }
             Frame::BGR0(frame) => {
                 println!(
-                    "Received frame of width {} and height {}",
+                    "Received BGR0 frame of width {} and height {}",
                     frame.width, frame.height
                 );
+                println!("Frame data => {:?}", frame.data.len());
```
Comment (Pranav2612000) [inline scap/src/main.rs]: This too.

File: scap/src/main.rs [hunk 7]
```diff
@@ -89,6 +98,13 @@ fn main() {
                     "Recieved RGBx frame of width {} and height {}",
                     frame.width, frame.height
                 );
+                println!("Frame data => {:?}", frame.data.len());
```
Comment (Pranav2612000) [inline scap/src/main.rs]: +1

File: scap/src/main.rs [hunk 8]
```diff
@@ -62,22 +64,29 @@ fn main() {
...
             Frame::BGR0(frame) => {
                 println!(
-                    "Received frame of width {} and height {}",
+                    "Received BGR0 frame of width {} and height {}",
                     frame.width, frame.height
                 );
+                println!("Frame data => {:?}", frame.data.len());
+                let x =
+                    image::RgbaImage::from_raw(frame.width as u32, frame.height as u32, frame.data);
+                let img = x.unwrap();
```
Comment (Pranav2612000) [inline scap/src/main.rs]: Let's not add this here. We'll keep the main function easy to work with. 
We can add this under examples later

File: scap/src/main.rs [hunk 9]
```diff
@@ -89,6 +98,13 @@ fn main() {
                     "Recieved RGBx frame of width {} and height {}",
                     frame.width, frame.height
                 );
+                println!("Frame data => {:?}", frame.data.len());
+                let x =
+                    image::RgbaImage::from_raw(frame.width as u32, frame.height as u32, frame.data);
+                let img = x.unwrap();
```
Comment (Pranav2612000) [inline scap/src/main.rs]: Same as the above comment. Let's not add this here for now.

File: scap/src/main.rs [hunk 10]
```diff
@@ -36,12 +38,12 @@ fn main() {
         show_cursor: true,
```
Comment (Pranav2612000) [inline scap/src/main.rs]: Is there a specific reason we've updated the output_type, source_rect values in `options`. If not, let's keep the older ones.

File: scap/src/main.rs [hunk 11]
```diff
@@ -1,9 +1,11 @@
 // This program is just a testbed for the library itself
 // Refer to the lib.rs file for the actual implementation
 
+use std::path::PathBuf;
+
 use scap::{
     capturer::{CGPoint, CGRect, CGSize, Capturer, Options},
-    frame::Frame,
+    frame::{Frame, FrameType},
```
Comment (Pranav2612000) [inline scap/src/main.rs]: This and the path import may not be required. Refer comments below

File: scap/src/capturer/engine/win/mod.rs [hunk 12]
```diff
@@ -66,29 +139,115 @@ impl WindowsCaptureHandler for Capturer {
...
+    
+    ).unwrap();
 
     return WinStream {
         settings,
         capture_control: None,
     };
 }
+
+pub fn get_output_frame_size(options: &Options) -> [u32; 2] {
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs:164]: Do we not need the scale factor logic the way we have for mac? Is it possible to get the `scale` value in windows?

File: scap/src/capturer/engine/win/mod.rs [hunk 13]
```diff
@@ -66,29 +139,115 @@ impl WindowsCaptureHandler for Capturer {
...
+) -> WinStream {
     let settings = Settings::new(
         Monitor::primary().unwrap(),
         Some(true),
         None,
-        ColorFormat::Rgba8,
-        tx,
-    )
-    .unwrap();
+        ColorFormat::Bgra8,
```
Comment (Pranav2612000) [inline scap/src/capturer/engine/win/mod.rs]: Let's get this color format from the frameType the user passes in options.

Comment (RohanPunjani) [reply scap/src/capturer/engine/win/mod.rs]: Should I modify the options struct?

Comment (Pranav2612000) [reply scap/src/capturer/engine/win/mod.rs]: We can match for `optioms.FrameType` and return Bgra8 if it's BGRAFrame and Rgba8 if it's RGBAFrame.

Does that make sense?

Comment (RohanPunjani) [reply scap/src/capturer/engine/win/mod.rs]: Modifying the struct

File: scap/src/capturer/engine/win/mod.rs (hunk 12 above)
Comment (RohanPunjani) [reply scap/src/capturer/engine/win/mod.rs:164]: Added code to get the scale value, but not using scale anywhere as of now!

File: scap/src/capturer/engine/win/mod.rs (hunk 13 above)
Comment (RohanPunjani) [reply scap/src/capturer/engine/win/mod.rs]: Will need some help here @Pranav2612000

**Pull request merged**